core/src/chain/header.rs

```diff
@@ -33,34 +38,29 @@ impl PartialEq for Header {
 }
 
 impl Header {
-    /// build a new Header from a chain, entry type and entry.
-    /// a Header is immutable, but the chain is mutable if chain.push() is used.
-    /// this means that a header becomes invalid and useless as soon as the chain is mutated
-    /// the only valid usage of a header is to immediately push it onto a chain in a Pair.
-    /// normally (outside unit tests) the generation of valid headers is internal to the
-    /// chain::SourceChain trait and should not need to be handled manually
-    ///
-    /// @see chain::pair::Pair
-    /// @see chain::entry::Entry
-    pub fn new(chain: &Chain, entry: &Entry) -> Header {
+    //
+    pub fn new(
+        entry_type: &str,
+        timestamp: &str,
+        link: Option<HashString>,
+        entry_hash: &str,
+        entry_signature: &str,
+        link_same_type: Option<HashString>,
+    ) -> Self {
         Header {
-            entry_type: entry.entry_type().clone(),
-            // @TODO implement timestamps
-            // https://github.com/holochain/holochain-rust/issues/70
-            timestamp: String::new(),
-            link: chain.top_pair().as_ref().map(|p| p.header().hash()),
-            entry_hash: entry.hash().to_string(),
-            link_same_type: chain
-                .top_pair_type(&entry.entry_type())
-                // @TODO inappropriate expect()?
-                // @see https://github.com/holochain/holochain-rust/issues/147
-                .map(|p| p.header().hash()),
-            // @TODO implement signatures
-            // https://github.com/holochain/holochain-rust/issues/71
-            entry_signature: String::new(),
+            entry_type: entry_type.to_string(),
+            timestamp: timestamp.to_string(),
+            link: link,
+            entry_hash: entry_hash.to_string(),
+            entry_signature: entry_signature.to_string(),
+            link_same_type: link_same_type,
         }
     }
 
+    pub fn new_from_json(header_str: &str) -> serde_json::Result<Self> {
```

Comment:
Better name this `from_json_str`. The `new_` is implicit when calling `Header::from_json_str("..")`.

Reply:
I was following pattern in Dna but I'll just change it everywhere then.